README.md
# angular2-meetup
Sample code for the JS meetup

# Warning

Angular 2 is under development. The API will change and this repository will be
obsolete withing a few days (written on May 10th, 2015).

# See this example running:

[http://andresdominguez.github.io/angular2-meetup/](http://andresdominguez.github.io/angular2-meetup/)

# Slides

[See the slides](https://docs.google.com/presentation/d/1ruBhy7DC2PmF-r9wHYPxMtkVSMOqIxcIBVR_nkCqJOo/edit?usp=sharing)

# Instructions

#### Clone this repository

```shell
$ git clone https://github.com/andresdominguez/angular2-meetup.git
```

#### Get TypeScript 1.5 beta

```shell
$ cd angular2-meetup
$ npm install -g typescript@^1.5.0-beta
$ tsc -w
message TS6042: Compilation complete. Watching for file changes.
```

#### Launch the server

Open a new tab, get the `http-server` through npm, and start the server.

```shell
$ npm install -g http-server
$ http-server
Starting up http-server, serving ./ on: http://0.0.0.0:8080
Hit CTRL-C to stop the server
```

Go to [http://localhost:8080/](http://localhost:8080/)


CHANGELOG.md
# July 3rd

* Upgrade to Angular 2 beta 28


app.ts
/// <reference path="typings/angular2/angular2.d.ts" />
import {Component, EventEmitter, NgFor, View, bootstrap} from 'angular2/angular2';
import {Http, httpInjectables} from "angular2/http";

interface HourlyForecast {
  c: string;
  f: string;
  time: string;
  day: string;
  clouds: number;
}

@Component({
  selector: 'city-selector',
  events: ['cityselected']
})
@View({
  template: `
    <div class="city-selector">
      <div class="city-selector-label">Choose a city: {{currentCity}}</div>
      <select name="city" id="city" class="form-control"
          (change)="cityChanged($event)">
        <option *ng-for="#city of cities"
            [selected]="city == currentCity">{{city}}</option>
      </select>
    </div>
  `,
  directives: [NgFor]
})
class CitySelector {
  cities: string[];
  currentCity: string;
  cityselected: any;

  constructor() {
    this.cities = [
      'Berlin',
      'Buenos Aires',
      'Cairo',
      'New York',
      'Sydney',
      'Tokyo'
    ];
    this.currentCity = 'New York';
    this.cityselected = new EventEmitter();
  }

  cityChanged(event) {
    this.currentCity = event.target.value;
    this.cityselected.next();
  }
}

@Component({
  selector: 'weather-card',
  properties: ['forecast']
})
@View({
  template: `
    <div class="weather-card">
      <div>{{forecast.day}} {{forecast.time}}</div>
      <div>
        <i class="wi" [class.wi-day-cloudy]="isCloudy()"
            [class.wi-day-sunny]="isSunny()">
          {{forecast.c}}C / {{forecast.f}}F
        </i>
      </div>
    </div>
  `
})
class WeatherCard {
  forecast: HourlyForecast;

  constructor() {
    this.forecast = null;
  }

  isCloudy(): boolean {
    return this.forecast.clouds > 30;
  }

  isSunny(): boolean {
    return this.forecast.clouds <= 30;
  }
}

class DateAndTime {
  static dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

  localTime: string;
  day: string;

  constructor(utcDate: string) {
    var date = new Date(utcDate + ' UTC');
    // Drop the seconds.
    this.localTime = date.toLocaleTimeString().replace(/:00 /, ' ');
    this.day = DateAndTime.dayNames[date.getDay()];
  }
}

@Component({
  selector: 'weather-app'
})
@View({
  template: `
    <div class="weather-app">
      <div>
        <div class="weather-app-city">{{cityName}}</div>
        <city-selector #city
            (cityselected)="cityChanged(city.currentCity)"></city-selector>
      </div>
      <div class="row">
        <div *ng-for="#item of forecastList" class="col-xs-4">
          <weather-card [forecast]="item"></weather-card>
        </div>
      </div>
    </div>
  `,
  directives: [CitySelector, NgFor, WeatherCard]
})
class WeatherApp {
  baseUrl: string;
  forecastList: HourlyForecast[];
  cityName: string;
  http: Http;

  constructor(http: Http) {
    this.http = http;
    this.baseUrl = 'http://api.openweathermap.org/data/2.5/forecast?q=';
    this.getWeather('New York');
  }

  cityChanged(currentCity: string) {
    this.getWeather(currentCity);
  }

  format(temperature: number): string {
    var stringValue = '' + temperature;
    // Get one digit after the decimal point.
    var index = stringValue.indexOf('.');
    if (index !== -1) {
      stringValue = stringValue.substring(0, index + 2);
    }
    return stringValue;
  }

  toCelsius(kelvin: string): string {
    // Temperature, Kelvin (subtract 273.15 to convert to Celsius)
    var celsius = parseFloat(kelvin) - 273.15;
    return this.format(celsius);
  }

  toFahrenheit(celsius: string): string {
    var fahrenheit = (parseFloat(celsius) * 1.8) + 32.0;
    return this.format(fahrenheit);
  }

  getWeather(currentCity: string) {
    this.http.get(this.baseUrl + currentCity)
        .map(response => response.json())
        .subscribe(response => this.handleResponse(response));
  }

  handleResponse(weatherData: any) {
    this.cityName = weatherData.city.name;

    this.forecastList = weatherData.list.map(item => {
      var c = this.toCelsius(item.main.temp);
      var dateAndTime = new DateAndTime(item.dt_txt);

      return {
        c: c,
        f: this.toFahrenheit(c),
        day: dateAndTime.day,
        time: dateAndTime.localTime,
        clouds: parseInt(item.clouds.all, 10)
      };
    });
  }
}

bootstrap(WeatherApp, [httpInjectables]);


app.js
if (typeof __decorate !== "function") __decorate = function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
if (typeof __metadata !== "function") __metadata = function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
/// <reference path="typings/angular2/angular2.d.ts" />
var angular2_1 = require('angular2/angular2');
var http_1 = require("angular2/http");
var CitySelector = (function () {
    function CitySelector() {
        this.cities = [
            'Berlin',
            'Buenos Aires',
            'Cairo',
            'New York',
            'Sydney',
            'Tokyo'
        ];
        this.currentCity = 'New York';
        this.cityselected = new angular2_1.EventEmitter();
    }
    CitySelector.prototype.cityChanged = function (event) {
        this.currentCity = event.target.value;
        this.cityselected.next();
    };
    CitySelector = __decorate([
        angular2_1.Component({
            selector: 'city-selector',
            events: ['cityselected']
        }),
        angular2_1.View({
            template: "\n    <div class=\"city-selector\">\n      <div class=\"city-selector-label\">Choose a city: {{currentCity}}</div>\n      <select name=\"city\" id=\"city\" class=\"form-control\"\n          (change)=\"cityChanged($event)\">\n        <option *ng-for=\"#city of cities\"\n            [selected]=\"city == currentCity\">{{city}}</option>\n      </select>\n    </div>\n  ",
            directives: [angular2_1.NgFor]
        }), 
        __metadata('design:paramtypes', [])
    ], CitySelector);
    return CitySelector;
})();
var WeatherCard = (function () {
    function WeatherCard() {
        this.forecast = null;
    }
    WeatherCard.prototype.isCloudy = function () {
        return this.forecast.clouds > 30;
    };
    WeatherCard.prototype.isSunny = function () {
        return this.forecast.clouds <= 30;
    };
    WeatherCard = __decorate([
        angular2_1.Component({
            selector: 'weather-card',
            properties: ['forecast']
        }),
        angular2_1.View({
            template: "\n    <div class=\"weather-card\">\n      <div>{{forecast.day}} {{forecast.time}}</div>\n      <div>\n        <i class=\"wi\" [class.wi-day-cloudy]=\"isCloudy()\"\n            [class.wi-day-sunny]=\"isSunny()\">\n          {{forecast.c}}C / {{forecast.f}}F\n        </i>\n      </div>\n    </div>\n  "
        }), 
        __metadata('design:paramtypes', [])
    ], WeatherCard);
    return WeatherCard;
})();
var DateAndTime = (function () {
    function DateAndTime(utcDate) {
        var date = new Date(utcDate + ' UTC');
        this.localTime = date.toLocaleTimeString().replace(/:00 /, ' ');
        this.day = DateAndTime.dayNames[date.getDay()];
    }
    DateAndTime.dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
    return DateAndTime;
})();
var WeatherApp = (function () {
    function WeatherApp(http) {
        this.http = http;
        this.baseUrl = 'http://api.openweathermap.org/data/2.5/forecast?q=';
        this.getWeather('New York');
    }
    WeatherApp.prototype.cityChanged = function (currentCity) {
        this.getWeather(currentCity);
    };
    WeatherApp.prototype.format = function (temperature) {
        var stringValue = '' + temperature;
        var index = stringValue.indexOf('.');
        if (index !== -1) {
            stringValue = stringValue.substring(0, index + 2);
        }
        return stringValue;
    };
    WeatherApp.prototype.toCelsius = function (kelvin) {
        var celsius = parseFloat(kelvin) - 273.15;
        return this.format(celsius);
    };
    WeatherApp.prototype.toFahrenheit = function (celsius) {
        var fahrenheit = (parseFloat(celsius) * 1.8) + 32.0;
        return this.format(fahrenheit);
    };
    WeatherApp.prototype.getWeather = function (currentCity) {
        var _this = this;
        this.http.get(this.baseUrl + currentCity)
            .map(function (response) { return response.json(); })
            .subscribe(function (response) { return _this.handleResponse(response); });
    };
    WeatherApp.prototype.handleResponse = function (weatherData) {
        var _this = this;
        this.cityName = weatherData.city.name;
        this.forecastList = weatherData.list.map(function (item) {
            var c = _this.toCelsius(item.main.temp);
            var dateAndTime = new DateAndTime(item.dt_txt);
            return {
                c: c,
                f: _this.toFahrenheit(c),
                day: dateAndTime.day,
                time: dateAndTime.localTime,
                clouds: parseInt(item.clouds.all, 10)
            };
        });
    };
    WeatherApp = __decorate([
        angular2_1.Component({
            selector: 'weather-app'
        }),
        angular2_1.View({
            template: "\n    <div class=\"weather-app\">\n      <div>\n        <div class=\"weather-app-city\">{{cityName}}</div>\n        <city-selector #city\n            (cityselected)=\"cityChanged(city.currentCity)\"></city-selector>\n      </div>\n      <div class=\"row\">\n        <div *ng-for=\"#item of forecastList\" class=\"col-xs-4\">\n          <weather-card [forecast]=\"item\"></weather-card>\n        </div>\n      </div>\n    </div>\n  ",
            directives: [CitySelector, angular2_1.NgFor, WeatherCard]
        }), 
        __metadata('design:paramtypes', [http_1.Http])
    ], WeatherApp);
    return WeatherApp;
})();
angular2_1.bootstrap(WeatherApp, [http_1.httpInjectables]);
